File: simulated_framework_playwright_ts/src/pages/BankingLoginPage.ts
/**
 * BankingLoginPage — playwright-typescript framework
 *
 * Page Object for the Banking Application Login page.
 * Requirement: REQ-BANK-AUTH-001
 *
 * Traceability:
 *   Requirement: REQ-BANK-AUTH-001
 *   Test Cases:  TC_001 – TC_008 (Banking_Login_TC_1.pdf)
 *   Framework:   playwright-typescript
 *   Rule refs:   PT-004, PT-005, PT-006, PT-008
 */

import { Page, Locator } from '@playwright/test';
import { BasePage } from './BasePage';

export class BankingLoginPage extends BasePage {
  // ── Locators ─────────────────────────────────────────────────────────────────
  private readonly pageTitle: Locator;
  private readonly usernameField: Locator;
  private readonly passwordField: Locator;
  private readonly loginButton: Locator;
  private readonly errorBanner: Locator;
  private readonly usernameValidationMsg: Locator;
  private readonly passwordValidationMsg: Locator;
  private readonly otpField: Locator;
  private readonly submitOtpButton: Locator;
  private readonly otpPage: Locator;
  private readonly forgotPasswordLink: Locator;
  private readonly rememberDeviceCheckbox: Locator;
  private readonly sessionExpiredBanner: Locator;

  constructor(page: Page) {
    super(page);
    this.pageTitle              = this.page.locator('h1.banking-login-title');
    this.usernameField          = this.page.locator('#bankingUsername');
    this.passwordField          = this.page.locator('#bankingPassword');
    this.loginButton            = this.page.locator('#bankingLoginBtn');
    this.errorBanner            = this.page.locator('div.login-error-banner');
    // Inline field-level validation messages (TC_006, TC_007, TC_008)
    this.usernameValidationMsg  = this.page.locator('#bankingUsername ~ span.field-error, #bankingUsername-error');
    this.passwordValidationMsg  = this.page.locator('#bankingPassword ~ span.field-error, #bankingPassword-error');
    // OTP page elements (TC_002)
    this.otpField               = this.page.locator('#otpInput');
    this.submitOtpButton        = this.page.locator('#submitOtpBtn');
    this.otpPage                = this.page.locator('div.otp-verification-page');
    this.forgotPasswordLink     = this.page.locator("a[data-testid='forgot-password']");
    this.rememberDeviceCheckbox = this.page.locator('#rememberDevice');
    this.sessionExpiredBanner   = this.page.locator('div.session-expired-banner');
  }

  // ── Validation ───────────────────────────────────────────────────────────────

  /**
   * Validates that the Banking Login page is displayed.
   */
  async validate(): Promise<boolean> {
    try {
      await this.waitForVisible(this.usernameField);
      await this.waitForVisible(this.loginButton);
      return true;
    } catch {
      return false;
    }
  }

  /**
   * Validates that the OTP verification page is displayed after first-factor login.
   * Used by TC_002.
   */
  async validateOtpPage(): Promise<boolean> {
    try {
      await this.waitForVisible(this.otpPage);
      await this.waitForVisible(this.otpField);
      return true;
    } catch {
      return false;
    }
  }

  // ── Actions ──────────────────────────────────────────────────────────────────

  async enterUsername(username: string): Promise<void> {
    await this.fillField(this.usernameField, username);
  }

  async enterPassword(password: string): Promise<void> {
    await this.fillField(this.passwordField, password);
  }

  async clickLoginButton(): Promise<void> {
    await this.clickWhenReady(this.loginButton);
  }

  /**
   * Clears the username field without entering any value, then moves focus away.
   * Used by TC_006 (empty username) and TC_008 (both fields empty).
   */
  async clearUsername(): Promise<void> {
    await this.waitForVisible(this.usernameField);
    await this.usernameField.clear();
  }

  /**
   * Clears the password field without entering any value, then moves focus away.
   * Used by TC_007 (empty password) and TC_008 (both fields empty).
   */
  async clearPassword(): Promise<void> {
    await this.waitForVisible(this.passwordField);
    await this.passwordField.clear();
  }

  /**
   * Performs a complete banking login (first factor only).
   * @param username - Account username / customer ID
   * @param password - Account password
   */
  async login(username: string, password: string): Promise<void> {
    await this.enterUsername(username);
    await this.enterPassword(password);
    await this.clickLoginButton();
  }

  /**
   * Enters valid first-factor credentials then clicks Login — used as the
   * precondition step for OTP flow in TC_002.
   */
  async loginFirstFactor(username: string, password: string): Promise<void> {
    await this.enterUsername(username);
    await this.enterPassword(password);
    await this.clickLoginButton();
  }

  /**
   * Submits the OTP code for two-factor authentication (TC_002).
   * @param otp - One-time passcode
   */
  async submitOtp(otp: string): Promise<void> {
    await this.fillField(this.otpField, otp);
    await this.clickWhenReady(this.submitOtpButton);
  }

  async clickForgotPasswordLink(): Promise<void> {
    await this.clickWhenReady(this.forgotPasswordLink);
  }

  async checkRememberDevice(): Promise<void> {
    await this.waitForVisible(this.rememberDeviceCheckbox);
    const checked = await this.rememberDeviceCheckbox.isChecked();
    if (!checked) {
      await this.rememberDeviceCheckbox.check();
    }
  }

  // ── Getters / Verifiers ──────────────────────────────────────────────────────

  async getErrorBannerText(): Promise<string> {
    await this.waitForVisible(this.errorBanner);
    return this.getText(this.errorBanner);
  }

  /**
   * Returns the inline validation message text for the username field.
   * Used by TC_006 and TC_008.
   */
  async getUsernameValidationText(): Promise<string> {
    await this.waitForVisible(this.usernameValidationMsg);
    return this.getText(this.usernameValidationMsg);
  }

  /**
   * Returns the inline validation message text for the password field.
   * Used by TC_007 and TC_008.
   */
  async getPasswordValidationText(): Promise<string> {
    await this.waitForVisible(this.passwordValidationMsg);
    return this.getText(this.passwordValidationMsg);
  }

  async isSessionExpiredBannerVisible(): Promise<boolean> {
    try {
      await this.waitForVisible(this.sessionExpiredBanner, 3_000);
      return true;
    } catch {
      return false;
    }
  }

  // ── Locator accessors for direct assertions ───────────────────────────────────
  getErrorBannerLocator(): Locator          { return this.errorBanner; }
  getLoginButtonLocator(): Locator          { return this.loginButton; }
  getUsernameFieldLocator(): Locator        { return this.usernameField; }
  getPasswordFieldLocator(): Locator        { return this.passwordField; }
  getUsernameValidationLocator(): Locator   { return this.usernameValidationMsg; }
  getPasswordValidationLocator(): Locator   { return this.passwordValidationMsg; }
  getOtpFieldLocator(): Locator             { return this.otpField; }
  getOtpPageLocator(): Locator              { return this.otpPage; }
}

File: simulated_framework_playwright_java/src/main/java/com/aether/pages/banking/BankingLoginPage.java
// ════════════════════════════════════════════════════
// SCRIPT READINESS: PARTIAL
// This script is structurally complete but contains
// placeholders for the following missing values.
// Supply these before executing:
//
// AUT_LOGIN_PATH  → set AUT_LOGIN_PATH in .env (assumed: /login)
// AUT_USER_EMAIL  → set AUT_USER_EMAIL in .env (assumed: process.env.AUT_USER_EMAIL)
// AUT_USER_PASSWORD → set AUT_USER_PASSWORD in .env (assumed: process.env.AUT_USER_PASSWORD)
// TARGET_BROWSER  → set TARGET_BROWSER in .env (assumed: framework primary browser)
// AUT_MODULES_IN_SCOPE → set AUT_MODULES_IN_SCOPE in .env (assumed: inferred from test case document titles/tags)
// LIVE_AUT_ACCESSIBLE → set LIVE_AUT_ACCESSIBLE in .env (assumed: false)
// ════════════════════════════════════════════════════
package com.aether.pages.banking;

import com.aether.framework.core.BasePage;
import com.microsoft.playwright.Locator;
import com.microsoft.playwright.Page;

public class BankingLoginPage extends BasePage {

    private final Locator pageTitle;
    private final Locator usernameField;
    private final Locator passwordField;
    private final Locator loginButton;
    private final Locator errorBanner;
    private final Locator usernameError;
    private final Locator passwordError;
    private final Locator otpField;
    private final Locator submitOtpButton;
    private final Locator otpPage;
    private final Locator forgotPasswordLink;
    private final Locator rememberDeviceCheckbox;
    private final Locator sessionExpiredBanner;

    public BankingLoginPage(Page page) {
        super(page);
        this.pageTitle              = page.locator("h1.banking-login-title");
        this.usernameField          = page.locator("#bankingUsername");
        this.passwordField          = page.locator("#bankingPassword");
        this.loginButton            = page.locator("#bankingLoginBtn");
        this.errorBanner            = page.locator("div.login-error-banner");
        this.usernameError          = page.locator("#bankingUsername-error, #bankingUsername ~ span.field-error");
        this.passwordError          = page.locator("#bankingPassword-error, #bankingPassword ~ span.field-error");
        this.otpField               = page.locator("#otpInput");
        this.submitOtpButton        = page.locator("#submitOtpBtn");
        this.otpPage                = page.locator("div.otp-verification-page, h2:has-text('Verify'), h2:has-text('OTP')").first();
        this.forgotPasswordLink     = page.locator("a[data-testid='forgot-password']");
        this.rememberDeviceCheckbox = page.locator("#rememberDevice");
        this.sessionExpiredBanner   = page.locator("div.session-expired-banner");
    }

    public boolean validate() {
        try {
            waitForVisible(usernameField);
            waitForVisible(loginButton);
            return true;
        } catch (Exception e) {
            return false;
        }
    }

    public boolean isStillOnLoginPage() {
        return validate();
    }

    public boolean validateOtpPage() {
        try {
            waitForVisible(otpPage);
            waitForVisible(otpField);
            return true;
        } catch (Exception e) {
            return false;
        }
    }

    public void enterUsername(String username) {
        fillField(usernameField, username);
    }

    public void enterPassword(String password) {
        fillField(passwordField, password);
    }

    public void clickLoginButton() {
        clickWhenReady(loginButton);
    }

    public void clearUsername() {
        waitForVisible(usernameField).clear();
    }

    public void clearPassword() {
        waitForVisible(passwordField).clear();
    }

    public void login(String username, String password) {
        enterUsername(username);
        enterPassword(password);
        clickLoginButton();
    }

    public void loginFirstFactor(String username, String password) {
        login(username, password);
    }

    public void submitOtp(String otp) {
        fillField(otpField, otp);
        clickWhenReady(submitOtpButton);
    }

    public boolean isErrorBannerDisplayed() {
        try {
            waitForVisible(errorBanner, 5_000);
            return true;
        } catch (Exception e) {
            return false;
        }
    }

    public String getErrorBannerText() {
        return getText(errorBanner);
    }

    public boolean isUsernameErrorDisplayed() {
        return isVisible(usernameError);
    }

    public boolean isPasswordErrorDisplayed() {
        return isVisible(passwordError);
    }

    public String getUsernameErrorText() {
        try {
            return getText(usernameError);
        } catch (Exception e) {
            return "";
        }
    }

    public String getPasswordErrorText() {
        try {
            return getText(passwordError);
        } catch (Exception e) {
            return "";
        }
    }

    public boolean isSessionExpiredBannerVisible() {
        try {
            waitForVisible(sessionExpiredBanner, 3_000);
            return true;
        } catch (Exception e) {
            return false;
        }
    }

    public Locator getUsernameFieldLocator() {
        return usernameField;
    }

    public Locator getPasswordFieldLocator() {
        return passwordField;
    }

    public Locator getLoginButtonLocator() {
        return loginButton;
    }

    public Locator getErrorBannerLocator() {
        return errorBanner;
    }

    public Locator getUsernameErrorLocator() {
        return usernameError;
    }

    public Locator getPasswordErrorLocator() {
        return passwordError;
    }

    public Locator getOtpFieldLocator() {
        return otpField;
    }

    public Locator getOtpPageLocator() {
        return otpPage;
    }

    public void openLoginPage() {
        // TODO(stub_and_continue): implement — resolve AUT_LOGIN_PATH and navigation strategy.
    }
}